src/java_bytecode/generate_java_generic_method.cpp

```diff
@@ -0,0 +1,296 @@
+    // is check that the proposed symbol name doesn't
+    // already exist.
+    throw "specialized method symbol already exists";
+  }
+
+  return *symbol_table.lookup(specialized_method.name);
+}
+
+
+/// Given a mapping from generic type variables to concrete types, generate
+/// a string suitable for decorating symbol names with.
+/// \param concrete_types The mapping from type variables to concrete types
+/// \return A string suitable for inserting into a java signature
+///
+/// As an example, if a mapping such as this (pseudo code...):
+/// [ {T,java.lang.Integer}, {U,java.lang.Double} ] is passed in, the return
+/// string would look like "<java.lang.Integer,java.lang,Double>"
+const std::string generate_java_generic_methodt::instantiation_decoration
+  (type_variable_instantiationst concrete_types) const
+{
+  std::ostringstream decorated_signature_buffer;
+  decorated_signature_buffer << "<";
+  bool first=true;
+  for(auto &concrete_type_entry : concrete_types)
+  {
+    if(first)
+      first=false;
+    else
+      decorated_signature_buffer << ",";
+
+    decorated_signature_buffer <<
+      concrete_type_entry.second.get(ID_identifier);
+  }
+  decorated_signature_buffer << ">";
+  return decorated_signature_buffer.str();
+}
+
+const void generate_java_generic_methodt::instantiate_java_generic_parameter(
+  java_generic_parametert &generic_parameter,
+  const generate_java_generic_methodt::type_variable_instantiation_mapt
+    &concrete_types) const
+{
+  INVARIANT(generic_parameter.id()==ID_pointer,
+            "All generic parameters should be pointers in java");
+  INVARIANT(generic_parameter.subtype().id()==ID_symbol,
+            "All generic parameters should point to symbols");
+
+  const symbol_typet &generic_type_variable=generic_parameter
+    .type_variable();
+  const auto &instantiation_type=concrete_types.find
+    (generic_type_variable);
+  if(instantiation_type==concrete_types.end())
+  {
+    // If we ever want/need to support partially instantiated types,
+    // we'll probably want to just return here.
+    throw
+      "No concrete type supplied for generic type variable in parameter";
+  }
+  generic_parameter.subtype()=instantiation_type->second;
+  generic_parameter.remove(ID_C_java_generic_parameter);
+  generic_parameter.remove(ID_type_variables);
+}
+
+
+/// Instantiate a given java generic type using a supplied mapping from type
+/// variable to concrete type.
+/// \param generic_type The java_generic_typet that should be modified and
+/// instantiated.
+/// \param concrete_types The mapping of type variables to concrete types.
+const void generate_java_generic_methodt::instantiate_java_generic_type(
+  java_generic_typet &generic_type,
+  const generate_java_generic_methodt::type_variable_instantiation_mapt
+    &concrete_types) const
+{
+  INVARIANT(generic_type.id()==ID_pointer,
+            "All generic parameters should be pointers in java");
+  INVARIANT(generic_type.subtype().id()==ID_symbol,
```

Comment:
Do you need this invariant? I think it isn't always true, sometimes we have a pointer to a struct. I think you can safely say:
```C++
ns.follow(generic_type.subtype()).id()==ID_struct
```

This is after resolving the symbol type you end up with a struct type (where `ns` is a namespacet` constructed from a `symbol_tablet`)

Reply:
I'd be curious to know under what conditions this `java_generic_typet` might be a pointer to a struct, as that might affect the correctness of the body of this function - but I'll happily update the invariant.